server/business-process-query/get-status-history-date-map.js

```diff
@@ -0,0 +1,420 @@
+'use strict';
+
+var oForEach    = require('es5-ext/object/for-each')
+  , debugLoad   = require('debug-ext')('load', 6)
+  , humanize    = require('debug-ext').humanize
+  , memoize     = require('memoizee')
+  , env         = require('mano').env
+  , filterSteps = require('./steps/filter')
+  , toDateInTz  = require('../../utils/to-date-in-time-zone')
+  , db          = require('../../db');
+
+/*
+Date map layout:
+
+[date][serviceName] = {
+	// Business process statuses: draft, revision, sentBack, process, pickup, rejected,
+	// withdrawn, closed
+	businessProcess: {
+		// 1.1 From [isSubmitted || revision || process] until ![revision || process]
+		pending: [bpId1, bpId2, …, bpIdn]
+		// 1.2 From pickup until !pickup
+		pickup: [bpId1, bpId2, …, bpIdn],
+		// 1.3 From sentBack until !sentBack
+		sentBack: [bpId1, bpId2, …, bpIdn],
+		// 1.4 Maps to isSubmitted
+		submitted: sum,
+		// 1.5 At withdrawn or closed
+		withdrawn: sum,
+		// 1.6 At rejected
+		rejected: sum
+	},
+	// Certificate statuses: pending, rejected, approved
+	certificate[certificateName]: {
+		// 2.1 From pending until [rejected || approved]
+		pending: [bpId1, bpId2, …, bpIdn]
+		// 2.2 On business process - From pickup until !pickup
+		pickup: [bpId1, bpId2, …, bpIdn],
+		// 2.3 On business process - From sentBack until !sentBack
+		sentBack: [bpId1, bpId2, …, bpIdn],
+		// 2.4 On business process - Maps to isSubmitted
+		submitted: sum,
+		// 2.5 On business process - At withdrawn or closed
+		withdrawn: sum,
+		// 2.6 On business process - At rejected
+		rejected: sum
+	},
+	// Processing step stauses: pending, paused, sentBack, rejected, approved, redelegated
+	processingStep[stepName]: {
+		// 3.1 From pending until !pending
+		pending: [bpId1, bpId2, …, bpIdn],
+		byProcessor[processorId]: {
+			// 3.2 From paused until !paused
+			paused: [bpId1, bpId2, …, bpIdn],
+			// 3.3 From sentBack until !sentBack
+			sentBack: [bpId1, bpId2, …, bpIdn],
+			// 3.4 From redelegated until !redelegated
+			redelegated: [bpId1, bpId2, …, bpIdn],
+			// 3.5 At rejected
+			rejected: [bpId1, bpId2, …, bpIdn],
+			// 3.6 At approved
+			approved: [bpId1, bpId2, …, bpIdn]
+		}
+	}
+};
+ */
+
+var getStatusHistoryLogs = function (statusHistory) {
+	return Array.from(statusHistory.values()).sort(function (a, b) {
+		return a.date - b.date;
+	});
+};
+
+module.exports = memoize(function (data) {
```

Comment:
Idea of having all status histories at all times in process memory, and on 24 hours intervals calculating from it _dates map_ is not efficient. We should not be reckless with keeping data in memory which we do not need keep.

It was specified that data of status histories should be gathered once and discarded right after we calculate _dates map_. Then we should update data for current date on _dates map_ by observing events on storages.
Let's follow this.

if incrementing data in _dates map_ per incoming status history records seems for some reason difficult, or requires to write another sophisticated handler, there are possibly ways to have something reusable that could be used to build _date map_ same way for _past_ and _upcoming_ events.
We do not need to provide any data for day which hasn't ended, we may add it to _dates map_ after it's done. So e.g. we may have a dedicated function `generateDay` which will be called for all past dates at initialization, and then after every day ends we may call it for day that passed.

Also this module should rather get data on it's own and not be fed with it, so it should be more as:

```javascript
module.exports = getData(..)(function (data) {
 // generate and return dates map
});
```